contracts/apps/ownership/OwnershipApp.sol

```diff
@@ -1,59 +1,87 @@
+// OwnershipApp requires ActionsOrgan to be installed in DAO
 
 // At the moment OwnershipApp intercepts MiniMe hook events, if governance app
 // needs them, it has to have higher priority than ownership app
 
-contract OwnershipApp is Application, Controller, Requestor {
+contract OwnershipConstants {
+  bytes4 constant addTokenSig = bytes4(sha3('addToken(address,uint256,uint128,uint128)'));
+  bytes4 constant removeTokenSig = bytes4(sha3('removeToken(address)'));
+  bytes4 constant getTokenSig = bytes4(sha3('getToken(uint256)'));
+  bytes4 constant issueTokensSig = bytes4(sha3('issueTokens(address,uint256)'));
+  bytes4 constant grantTokensSig = bytes4(sha3('grantTokens(address,address,uint256)'));
+  bytes4 constant grantVestedTokensSig = bytes4(sha3('grantVestedTokens(address,address,uint256,uint64,uint64,uint64)'));
+}
+
+contract OwnershipApp is OwnershipConstants, Application, Controller, Requestor {
   struct Token {
     address tokenAddress;
     uint128 governanceRights;
     uint128 economicRights;
   }
 
   Token[] tokens;
+  mapping (address => uint) public tokenIdForAddress;
+
+  event AddedToken(address tokenAddress, uint tokenId);
+  event RemovedToken(address tokenAddress);
+  event ChangedTokenId(address tokenAddress, uint oldTokenId, uint newTokenId);
+
   uint8 constant maxTokens = 20; // prevent OOGs when tokens are iterated
   uint constant holderThreshold = 1; // if owns x tokens is considered holder
 
   function OwnershipApp(address daoAddr)
-           Application(daoAddr) {}
+           Application(daoAddr) {
+    tokens.push(Token(0,0,0)); // to prevent index 0
+  }
 
-  function addOrgToken(address tokenAddress, uint256 issueAmount, uint128 governanceRights, uint128 economicRights) onlyDAO {
+  function addToken(address tokenAddress, uint256 issueAmount, uint128 governanceRights, uint128 economicRights) onlyDAO {
     // Only add tokens the DAO is the controller of, so we can control it.
     // If it is a wrap over another token, the Wrap implementation can remove some functionality.
     require(MiniMeToken(tokenAddress).controller() == dao);
-    uint256 tokenId = TokensOrgan(dao).addToken(tokenAddress);
     uint newLength = tokens.push(Token(tokenAddress, governanceRights, economicRights));
-    assert(tokenId == newLength - 1 && newLength <= maxTokens);
-    if (issueAmount > 0) issueTokens(tokenId, issueAmount);
+    uint256 tokenId = newLength - 1;
+    tokenIdForAddress[tokenAddress] = tokenId;
+
+    AddedToken(tokenAddress, tokenId);
+
+    if (issueAmount > 0) issueTokens(tokenAddress, issueAmount);
   }
 
-  function removeOrgToken(uint256 tokenId) onlyDAO {
-    TokensOrgan(dao).removeToken(tokenId);
-    if (tokens.length > 1) tokens[tokenId] = tokens[tokens.length - 1];
+  function removeToken(address tokenAddress) onlyDAO {
+    uint tokenId = tokenIdForAddress[tokenAddress];
+    require(tokenId > 0);
+    if (tokens.length > 1) {
+      tokens[tokenId] = tokens[tokens.length - 1];
```

Comment:
The way that we do the remove very cheaply is to remove the last item of the array (with `tokens.length -= 1`) so the token at that index is replaced at the index of the token that is being deleted